### PR TITLE
Help compiler enforce nullability annotations

### PR DESCRIPTION
**Pull request by jcouv**
TL;DR: the goal for this PR is to confirm that some upcoming compiler changes for VS 16.6 are okay. ~~This PR is intended for discussion but not to merge as-is :-)~~

Tagging @stephentoub @safern for discussion.

----
Background:

I have been working on [compiler changes](https://github.com/dotnet/roslyn/issues/36039) which adds some warnings as a result of nullability attributes.

For instance:
- `[MaybeNull] string parameter` would start with a maybe-null state.
- `[MaybeNull] override T M()` cannot override `virtual T M()`
- we check the state of `[MaybeNullWhen(true)] string parameter`
- we check that methods with `[DoesNotReturn]` indeed don't return

Jared recommended that I test the change on the `runtime` repo, and it indeed produced more warnings. Here's a PR that addresses some of the issues flagged and tags some others.

Issues identified so far:
- some `GetHashCode` methods are marked as `[DisallowNull]` and that's causing problems
- `FailFast` is marked with `[DoesNotReturn]`, but it looks like it could return in some cases
- I've tagged a few places where methods with conditional `[Maybe/NotNullWhen(...)]` attributes seem incorrectly implemented

### CODE REVIEW

File: src/libraries/System.Collections.Immutable/src/System/Collections/Immutable/ImmutableHashSet_1.cs
Comment (jcouv) [inline src/libraries/System.Collections.Immutable/src/System/Collections/Immutable/ImmutableHashSet_1.cs]: :memo: should we remove the `[DisallowNull]` on that `GetHashCode`, or add `[DisallowNull]` here?
(same question below)

File: src/libraries/System.ComponentModel.Composition/src/System/ComponentModel/Composition/Primitives/ComposablePartDefinition.cs
Comment (jcouv) [inline src/libraries/System.ComponentModel.Composition/src/System/ComponentModel/Composition/Primitives/ComposablePartDefinition.cs]: :memo: the `[NotNullWhen(...)]` annotation on method seems incorrect

File: src/libraries/System.ComponentModel.Composition/src/System/ComponentModel/Composition/ReflectionModel/ImportType.cs
Comment (jcouv) [inline src/libraries/System.ComponentModel.Composition/src/System/ComponentModel/Composition/ReflectionModel/ImportType.cs]: :memo: the `[NotNullWhen(...)]` annotation on method seems incorrect

File: src/libraries/System.ComponentModel.Composition/src/System/ComponentModel/Composition/ReflectionModel/ReflectionComposablePartDefinition.cs
Comment (jcouv) [inline src/libraries/System.ComponentModel.Composition/src/System/ComponentModel/Composition/ReflectionModel/ReflectionComposablePartDefinition.cs]: :memo: the `[NotNullWhen(...)]` annotation on method seems incorrect

File: src/libraries/System.Linq.Expressions/src/System/Linq/Expressions/Interpreter/LightCompiler.cs
Comment (jcouv) [inline src/libraries/System.Linq.Expressions/src/System/Linq/Expressions/Interpreter/LightCompiler.cs]: :memo: the `[NotNullWhen(...)]` annotation on method seems incorrect

File: src/libraries/System.Linq.Expressions/src/System/Linq/Expressions/Interpreter/LightCompiler.cs
Comment (jcouv) [inline src/libraries/System.Linq.Expressions/src/System/Linq/Expressions/Interpreter/LightCompiler.cs]: :memo: `IComparer` declares that it can compare `null` inputs, but this implementation doesn't handle.

File: src/libraries/System.Private.CoreLib/src/System/Diagnostics/DebugProvider.cs
Comment (jcouv) [inline src/libraries/System.Private.CoreLib/src/System/Diagnostics/DebugProvider.cs]: :memo: without this suppression, the compiler points out that this method could return normally. Looking at the implementation of `FailCore`, I think it has the same problem.
We need to discuss what is the best way to suppress this warning. Maybe a more fined-grained suppression on a specific diagnostic would be better than the coarse `#nullable disable` I used...

File: src/libraries/System.Private.CoreLib/src/System/Runtime/InteropServices/MemoryMarshal.cs
Comment (jcouv) [inline src/libraries/System.Private.CoreLib/src/System/Runtime/InteropServices/MemoryMarshal.cs]: :memo: we could remove the `localManager` local.

File: src/libraries/System.Private.CoreLib/src/System/Runtime/InteropServices/MemoryMarshal.cs
Comment (jcouv) [inline src/libraries/System.Private.CoreLib/src/System/Runtime/InteropServices/MemoryMarshal.cs]: :memo: we could remove the `localManager` local.

File: src/libraries/System.Reflection.MetadataLoadContext/src/System/Reflection/TypeLoading/General/RoAssemblyName.cs
Comment (jcouv) [inline src/libraries/System.Reflection.MetadataLoadContext/src/System/Reflection/TypeLoading/General/RoAssemblyName.cs]: :memo: not sure what to do here

File: src/libraries/System.Runtime.Extensions/ref/System.Runtime.Extensions.cs
Comment (jcouv) [inline src/libraries/System.Runtime.Extensions/ref/System.Runtime.Extensions.cs]: :memo: I'm not sure why there isn't a warning on `throw null`. I would expect `throw null!` would be required. Is this suppressed for the file or something?

File: src/libraries/System.Collections.Immutable/src/System/Collections/Immutable/ImmutableHashSet_1.cs
Comment (stephentoub) [reply src/libraries/System.Collections.Immutable/src/System/Collections/Immutable/ImmutableHashSet_1.cs]: In general `IEqualityComparer<T>.GetHashCode` (e.g. StringComparer.GetHashCode throws ArgumentNullException for null) even if some implementations do, so we can't remove `[DisallowNull]`.  We could consider adding `[DisallowNull]` to this `T equalValue`, but then we'd be saying that null wasn't allowed even if the developer provided a comparer implementation that did actually allow null.  This is an example where we lack sufficient expressivity.  I guess for now it'd be best to add `[DisallowNull]`, and then see what kind of issues that causes; we can remove `[DisallowNull]` later on, but it'll be harder to add it later on.

File: src/libraries/System.Collections.Immutable/src/System/Collections/Immutable/ImmutableHashSet_1.cs
Comment (stephentoub) [inline src/libraries/System.Collections.Immutable/src/System/Collections/Immutable/ImmutableHashSet_1.cs]: But now it seems like that would actually not just be for TryGetValue, but really all operations on the hash set, at which point we'd be better off using a notnull constraint?  Maybe instead of changing the annotations, open an issue about it and keep the ! changes you have here.

Comment (stephentoub) [reply src/libraries/System.Collections.Immutable/src/System/Collections/Immutable/ImmutableHashSet_1.cs]: (And actually, the right answer is probably to do a null check prior to calling GetHashCode, as we do in `HashSet<T>`.)

File: src/libraries/System.ComponentModel.Composition/src/System/ComponentModel/Composition/Primitives/ComposablePartDefinition.cs
Comment (stephentoub) [reply src/libraries/System.ComponentModel.Composition/src/System/ComponentModel/Composition/Primitives/ComposablePartDefinition.cs]: Yes, please remove it.  Thanks.

File: src/libraries/System.ComponentModel.Composition/src/System/ComponentModel/Composition/ReflectionModel/ImportType.cs
Comment (stephentoub) [reply src/libraries/System.ComponentModel.Composition/src/System/ComponentModel/Composition/ReflectionModel/ImportType.cs]: Yes, please remove it.  The dangers of the compiler not having checked those annotations ;)

File: src/libraries/System.ComponentModel.Composition/src/System/ComponentModel/Composition/ReflectionModel/ReflectionComposablePartDefinition.cs
Comment (stephentoub) [reply src/libraries/System.ComponentModel.Composition/src/System/ComponentModel/Composition/ReflectionModel/ReflectionComposablePartDefinition.cs]: Ditto

File: src/libraries/System.Linq.Expressions/src/System/Linq/Expressions/Interpreter/LightCompiler.cs
Comment (stephentoub) [reply src/libraries/System.Linq.Expressions/src/System/Linq/Expressions/Interpreter/LightCompiler.cs]: See the comment above: "unreachable".

File: src/libraries/Common/src/System/Collections/Generic/ReferenceEqualityComparer.cs
Comment (safern) [inline src/libraries/Common/src/System/Collections/Generic/ReferenceEqualityComparer.cs]: Since `T` is constraint to class, can we instead annotate `T` as `T?` ?

File: src/libraries/System.Linq.Expressions/src/System/Linq/Expressions/Interpreter/LightCompiler.cs
Comment (stephentoub) [reply src/libraries/System.Linq.Expressions/src/System/Linq/Expressions/Interpreter/LightCompiler.cs]: It's only ever used with non-null values: https://source.dot.net/#System.Linq.Expressions/System/Linq/Expressions/Interpreter/LightCompiler.cs,233

File: src/libraries/System.Linq.Parallel/src/System/Linq/Parallel/QueryOperators/Unary/ForAllOperator.cs
Comment (stephentoub) [inline src/libraries/System.Linq.Parallel/src/System/Linq/Parallel/QueryOperators/Unary/ForAllOperator.cs]: Is this correct?  With the `[AllowNull]`, someone could pass in a null `TInput` even if `TInput` is declared to be non-nullable, in which case if the method doesn't touch `currentElement`, it'll still be null on the way out.  Doesn't that mean it should remain `[MaybeNull]`?

File: src/libraries/System.Linq.Expressions/src/System/Linq/Expressions/Interpreter/LightCompiler.cs
Comment (safern) [reply src/libraries/System.Linq.Expressions/src/System/Linq/Expressions/Interpreter/LightCompiler.cs]: Based on the comment this code block is unreachable.

File: src/libraries/System.Private.CoreLib/src/System/Diagnostics/DebugProvider.cs
Comment (stephentoub) [reply src/libraries/System.Private.CoreLib/src/System/Diagnostics/DebugProvider.cs]: You're nullable disabling/enabling around the closing brace?

Regardless, yes, it's possible for someone to implement Fail in a way that still returns, but that goes against the purpose of the method.  If we held true to the possibility that someone did that, then we'd need to remove `[DoesNotReturnIf(false)]` from Debug.Assert and all such methods, which would be terrible.

File: src/libraries/System.Private.CoreLib/src/System/HashCode.cs
Comment (stephentoub) [inline src/libraries/System.Private.CoreLib/src/System/HashCode.cs]: As with the earlier examples with ImmutableHashSet, we should probably fix this implementation.  e.g. today if you wrote code like:
```C#
#nullable enable
using System;

class Program
{
    static void Main(string[] args)
    {
        HashCode c = default;
        c.Add(null, StringComparer.Ordinal);
    }
}
```
it will blow up.  Seems like this line should instead be:
```C#
Add(comparer != null ? comparer.GetHashCode(value ?? 0) : (value?.GetHashCode() ?? 0));
```

File: src/libraries/System.Private.CoreLib/src/System/Diagnostics/DebugProvider.cs
Comment (safern) [reply src/libraries/System.Private.CoreLib/src/System/Diagnostics/DebugProvider.cs]: Can't we add `DoesNotReturn` to `FailCore`?

File: src/libraries/System.Reflection.MetadataLoadContext/src/System/Reflection/TypeLoading/General/RoAssemblyName.cs
Comment (safern) [reply src/libraries/System.Reflection.MetadataLoadContext/src/System/Reflection/TypeLoading/General/RoAssemblyName.cs]: why did we need to make it nullable?

File: src/libraries/System.Private.CoreLib/src/System/Runtime/InteropServices/MemoryMarshal.cs
Comment (stephentoub) [inline src/libraries/System.Private.CoreLib/src/System/Runtime/InteropServices/MemoryMarshal.cs]: What's wrong here?

File: src/libraries/System.Private.CoreLib/src/System/Runtime/InteropServices/MemoryMarshal.cs
Comment (stephentoub) [inline src/libraries/System.Private.CoreLib/src/System/Runtime/InteropServices/MemoryMarshal.cs]: What's wrong here?

File: src/libraries/System.Runtime.Extensions/ref/System.Runtime.Extensions.cs
Comment (safern) [reply src/libraries/System.Runtime.Extensions/ref/System.Runtime.Extensions.cs]: Yes, we suppress that for all reference assemblies. However, since this is autogenerated, this code will be removed by the autogenerated tool because the return type is `void`. 

We should fix that in our tool generation. I created: https://github.com/dotnet/arcade/issues/4820

File: src/libraries/System.Reflection.MetadataLoadContext/src/System/Reflection/TypeLoading/General/RoAssemblyName.cs
Comment (stephentoub) [reply src/libraries/System.Reflection.MetadataLoadContext/src/System/Reflection/TypeLoading/General/RoAssemblyName.cs]: This is an internal type, and presumably it's not expecting Equals to be used with null, even though it implements an interface that allows it.  Seems like the right change is to remove this ! and ? and suppress the warning on the Equals method signature.

File: src/libraries/System.Private.CoreLib/src/System/Diagnostics/DebugProvider.cs
Comment (jcouv) [reply src/libraries/System.Private.CoreLib/src/System/Diagnostics/DebugProvider.cs]: @safern  `FailCore` has `return` statements, so the same warning "method should not return" would be produced there if we add `[DoesNotReturn]`.

File: src/libraries/System.Reflection.MetadataLoadContext/src/System/Reflection/TypeLoading/General/RoAssemblyName.cs
Comment (jcouv) [reply src/libraries/System.Reflection.MetadataLoadContext/src/System/Reflection/TypeLoading/General/RoAssemblyName.cs]: To implement `IEquatable` you must be compatible with the signature `bool Equals([AllowNull] T other);`

File: src/libraries/System.Runtime.Extensions/ref/System.Runtime.Extensions.cs
Comment (stephentoub) [reply src/libraries/System.Runtime.Extensions/ref/System.Runtime.Extensions.cs]: Yes:
```
eng\referenceAssemblies.props:    <!-- disable CS8597 because we throw null on reference assemblies. -->
eng\referenceAssemblies.props:    <NoWarn>$(NoWarn);CS8597</NoWarn>
```

File: src/libraries/System.Reflection.MetadataLoadContext/src/System/Reflection/TypeLoading/General/RoAssemblyName.cs
Comment (stephentoub) [reply src/libraries/System.Reflection.MetadataLoadContext/src/System/Reflection/TypeLoading/General/RoAssemblyName.cs]: > To implement IEquatable you must be compatible with the signature bool Equals([AllowNull] T other);

I understand that.  So I'm saying use `#pragma warning disable whatever` to suppress the warning.

File: src/libraries/Common/src/System/Collections/Generic/ReferenceEqualityComparer.cs
Comment (jcouv) [reply src/libraries/Common/src/System/Collections/Generic/ReferenceEqualityComparer.cs]: Indeed. I'd missed that. Thanks

File: src/libraries/System.Linq.Parallel/src/System/Linq/Parallel/QueryOperators/Unary/ForAllOperator.cs
Comment (jcouv) [reply src/libraries/System.Linq.Parallel/src/System/Linq/Parallel/QueryOperators/Unary/ForAllOperator.cs]: `QueryOperatorEnumerator` declares `internal abstract bool MoveNext([MaybeNullWhen(false), AllowNull] ref TElement currentElement, ref TKey currentKey);`.

So either `QueryOperatorEnumerator.MoveNext` needs to change to `[MaybeNull]` or `ForAllEnumerator.MoveNext` does ;-)

Comment (stephentoub) [reply src/libraries/System.Linq.Parallel/src/System/Linq/Parallel/QueryOperators/Unary/ForAllOperator.cs]: This is all internal in code that hasn't changed in 10 years.  I have a fairly low care-level for the annotations; do whichever you think makes the most sense, as long as the generated IL for the actual method code doesn't change :)

File: src/libraries/System.Private.CoreLib/src/System/Runtime/InteropServices/MemoryMarshal.cs
Comment (jcouv) [reply src/libraries/System.Private.CoreLib/src/System/Runtime/InteropServices/MemoryMarshal.cs]: The compiler doesn't know that `manager == localManager`, so it looks like we may be returning with a `null` manager when this method returns `true`.
We could fix that by doing `return manager != null;`, but that defeats the purpose indicated by above comment ("Use register for null comparison rather than byref").

File: src/libraries/System.Private.CoreLib/src/System/Runtime/InteropServices/MemoryMarshal.cs
Comment (jcouv) [reply src/libraries/System.Private.CoreLib/src/System/Runtime/InteropServices/MemoryMarshal.cs]: Ditto

File: src/libraries/System.Private.CoreLib/src/System/Runtime/InteropServices/MemoryMarshal.cs
Comment (stephentoub) [reply src/libraries/System.Private.CoreLib/src/System/Runtime/InteropServices/MemoryMarshal.cs]: Ah, so this TODO was about improving the compiler :)

For the purposes of this change, just suppress the warning with whatever syntax is least obtrusive.

File: src/libraries/Common/src/System/Collections/Generic/ReferenceEqualityComparer.cs
Comment (jcouv) [reply src/libraries/Common/src/System/Collections/Generic/ReferenceEqualityComparer.cs]: Hum, it looks like this file is compiled into more than one project. One with nullability enabled and one without. Reverted to use the attribute.

Comment (stephentoub) [reply src/libraries/Common/src/System/Collections/Generic/ReferenceEqualityComparer.cs]: Can you use the annotation and just add `#nullable enable` at the top of the file as well?  That's what we've done in other cases like this, where we've not yet gotten to every project that includes it.

File: src/libraries/System.ComponentModel.Composition/src/Microsoft/Internal/Collections/CollectionServices.cs
Comment (stephentoub) [inline src/libraries/System.ComponentModel.Composition/src/Microsoft/Internal/Collections/CollectionServices.cs]:
```suggestion
                Debug.Assert(value != null);
```

Comment (stephentoub) [inline src/libraries/System.ComponentModel.Composition/src/Microsoft/Internal/Collections/CollectionServices.cs]: How do we know this assert is valid?

File: src/libraries/System.Linq.Expressions/src/System/Dynamic/Utils/CollectionExtensions.cs
Comment (stephentoub) [inline src/libraries/System.Linq.Expressions/src/System/Dynamic/Utils/CollectionExtensions.cs]:
```suggestion
                if (t != null)
```

File: src/libraries/System.Linq.Expressions/src/System/Linq/Expressions/Interpreter/LightCompiler.cs
Comment (stephentoub) [inline src/libraries/System.Linq.Expressions/src/System/Linq/Expressions/Interpreter/LightCompiler.cs]:
```suggestion
                Debug.Assert(d1 != null && d2 != null);
```

File: src/libraries/System.Linq.Parallel/src/System/Linq/Parallel/Utils/WrapperEqualityComparer.cs
Comment (stephentoub) [inline src/libraries/System.Linq.Parallel/src/System/Linq/Parallel/Utils/WrapperEqualityComparer.cs]:
```suggestion
            return value != null ? 0 : _comparer.GetHashCode(value);
```

File: src/libraries/System.Private.CoreLib/src/System/Action.cs
Comment (stephentoub) [inline src/libraries/System.Private.CoreLib/src/System/Action.cs]: Good catch.  We need to update the ref assembly accordingly:
https://github.com/dotnet/runtime/blob/ca6adaac796a19d70e4361a9b2a8d090c2f27637/src/libraries/System.Runtime/ref/System.Runtime.cs#L611

File: src/libraries/System.ComponentModel.Composition/src/Microsoft/Internal/Collections/CollectionServices.cs
Comment (jcouv) [reply src/libraries/System.ComponentModel.Composition/src/Microsoft/Internal/Collections/CollectionServices.cs]: The two callers get their values from `ComposablePartDefinition.TryGetExports`. That method can return with 
- both `singleMatch` and `multipleMatches` being null (`return false;`, but then `FastAppendToListAllowNulls` won't be called because of calling pattern, copied below), 
- or `multipleMatches` being set and returning true, 
- or `singleMatch` being set and returning true.

`singleMatch` corresponds to parameter `value` and `multipleMatches` corresponds to parameter `second` here.

Calling pattern:
``` csharp
                    if (part.TryGetExports(definition, out Tuple<ComposablePartDefinition, ExportDefinition>? singleMatch, out IEnumerable<Tuple<ComposablePartDefinition, ExportDefinition>>? multipleMatches))
                    {
                        exports = exports.FastAppendToListAllowNulls(singleMatch, multipleMatches);
                    }
```

---
In reply to: [378614963](https://github.com/dotnet/runtime/pull/32090#discussion_r378614963) [](ancestors = 378614963)

File: src/libraries/System.Private.CoreLib/src/System/Runtime/InteropServices/MemoryMarshal.cs
Comment (jcouv) [inline src/libraries/System.Private.CoreLib/src/System/Runtime/InteropServices/MemoryMarshal.cs]: >#pragma warning disable 8762  [](start = 0, length = 29)

:memo: commit 11 experimented with replacing pragma with `Debug.Assert`, but that results in different IL, so I reverted.

Comment (stephentoub) [reply src/libraries/System.Private.CoreLib/src/System/Runtime/InteropServices/MemoryMarshal.cs]: >but that results in different IL, so I reverted.

In release or just in debug?  Just in debug is fine :)

Comment (jcouv) [reply src/libraries/System.Private.CoreLib/src/System/Runtime/InteropServices/MemoryMarshal.cs]: Unfortunately the impact is in release.
`if (x == null) return true; else return false;` is different than `return x == null;`

[sharplab](https://sharplab.io/#v2:C4LglgNgPgAgTARgLACgYGYAE9MGFMDeqmJ2WARgPaUSYCyAFAMrABOYAdgOaYDOAlJmKkiKUuMxgAZpgZ9MAXgWYOAVwgR+wiYW06SMAOyY2qgKYBuUnvEBfG6SOYpAQwi9Le+yltA=)

It's possible that the runtime optimizes this away in assembly, but I didn't want to chance it...

Comment (stephentoub) [reply src/libraries/System.Private.CoreLib/src/System/Runtime/InteropServices/MemoryMarshal.cs]: Ah, thanks, no, let's not do that :-)

File: src/libraries/System.Text.Json/src/System/Text/Json/ThrowHelper.Serialization.cs
Comment (ahsonkhan) [inline src/libraries/System.Text.Json/src/System/Text/Json/ThrowHelper.Serialization.cs]: I am curious, how did you catch this typo?

Comment (jcouv) [reply src/libraries/System.Text.Json/src/System/Text/Json/ThrowHelper.Serialization.cs]: A new version of the compiler checks method bodies to see that they honor the nullability attributes that they declare in their API.  This method has `return` statements, so it was flagged ;-)

Comment (ahsonkhan) [reply src/libraries/System.Text.Json/src/System/Text/Json/ThrowHelper.Serialization.cs]: Nice!

**Comment by jcouv** [comment]
I believe this is ready to merge/squash if I can get the required sign-off(s). Thanks

**Comment by stephentoub** [comment]
Thanks, Julien.

**Review by stephentoub — APPROVED** [review]

**Comment by jcouv** [comment]
Thanks!

**Pull request merged**
